src/Symfony/Component/Security/Core/Authorization/Voter/Voter.php

```diff
@@ -27,25 +27,52 @@ abstract class Voter implements VoterInterface
     public function vote(TokenInterface $token, $subject, array $attributes)
     {
         // abstain vote by default in case none of the attributes are supported
-        $vote = self::ACCESS_ABSTAIN;
+        $vote = $this->abstain();
 
         foreach ($attributes as $attribute) {
             if (!$this->supports($attribute, $subject)) {
                 continue;
             }
 
             // as soon as at least one attribute is supported, default is to deny access
-            $vote = self::ACCESS_DENIED;
+            $vote = $this->deny();
 
-            if ($this->voteOnAttribute($attribute, $subject, $token)) {
+            $v = \is_bool($v = $this->voteOnAttribute($attribute, $subject, $token)) ? Vote::create($v) : $v; // BC layer
+            if ($v->isGranted()) {
                 // grant access as soon as at least one attribute returns a positive response
-                return self::ACCESS_GRANTED;
+                return $v;
+            } else {
+                $vote->merge($v);
             }
         }
 
         return $vote;
     }
 
+    /**
+     * Creates an granted vote.
```

Comment:
```suggestion
     * Creates a granted vote.
```

(I am not english native, but `an granted` seems weird)